kernel/safe_typing.ml

```diff
@@ -748,24 +747,32 @@ let inline_private_constants env ((body, ctx), side_eff) =
   let ctx' = Univ.ContextSet.union ctx ctx' in
   (body, ctx')
 
+let warn_failed_cert = CWarnings.create ~name:"failed-abstract-certificate"
+    ~category:CWarnings.CoreCategories.tactics ~default:CWarnings.Disabled
+    Pp.(fun kn ->
+        str "Certificate for private constant " ++
+        Label.print (Constant.label kn) ++
+        str " failed.")
```

Comment:
NB: trying to make this an error in stdlib and test suite fails on
- https://github.com/coq/coq/blob/63cf312cdebc40983a7c8273de1d9e6dc0600efe/theories/Numbers/Cyclic/Int63/Uint63.v#L884 (probably `auto with zarith` doing some `abstract` then backtracking)
- #13834 (`rewrite in *` makes a side effect then backtracks)
- #5065 (Qed mutual fixpoint needs to declare a side effect for each fixpoint (since Qed makes them inlined the declaration from the 1st fixpoint is not available to the 2nd so must be redeclared), but the certificate doesn't match the current senv when declaring the 2nd fixpoint
- #5641 (`abstract` in `ltac:()`)

Backtracking and `abstract` in `ltac:()` probably can't be fixed without refactoring side effects to live in the evar map instead of the global env.
Qed mutual fixpoints probably need some custom handling, not sure. Not much point worrying about it while we still have the backtracking issues.

Reply:
We can also get rid of side effects at all, the idea was nice but IMHO newer approaches are superior.

Reply:
IDK what newer approaches you're talking about.

Reply:
The approach is to stop guessing what proofs do (and trying to export it), instead use a simple linear model, skip the proof first, then if the proof is computed asynchronously, you compare the output with the output of skip.

If they are the same, you are good, otherwise you'll have to redo the document; I've added an entry to the call in case we wanna discuss about it.

For sure we need some stats of the effectiveness of the approach, but also the current approach actually performs bad in practice so I'm not sure why we want to keep it.

Reply:
It would mean we don't need to merge the univs in close proof, we just keep the linear semantics. But indeed I need to verify that, I'm a bit busy today, so take what I say as totally unchecked.

Last time I looked we had quite a few problems in that area and with the close_proof semantcis, so I am just wondering what is less effort: to recover a linear semantics, or to fix that model (which I'm very unconvinced it is worth it)

Reply:
It would mean we don't need to merge the univs in close proof, we just keep the linear semantics. But indeed I need to verify that, I'm a bit busy today, so take what I say as totally unchecked.

Last time I looked we had quite a few problems in that area and with the close_proof semantcis, so I am just wondering what is less effort: to recover a linear semantics, or to fix that model (which I'm very unconvinced it is worth it)